src/pages/Interferometry/index.js

```diff
@@ -1,4 +1,96 @@
-import React, { useState } from "react";
+import React, { useRef, useState } from "react";
+import useAnimationFrame from "../../useAnimation";
+
+class DoubleRect {
+  constructor(x, y, w = 50, h = 50, angle = 0, strokeRatio = 0, fillRatio = 1) {
+    this.x = x;
+    this.y = y;
+    this.w = w;
+    this.h = h;
+    this.angle = angle;
+    this.radian = this.angle * (Math.PI / 180);
+    this.strokeRatio = strokeRatio;
+    this.fillRatio = fillRatio;
+  }
+
+  draw(ctx) {
+    ctx.save();
+    ctx.translate(this.x, this.y);
+    ctx.rotate(this.radian);
+    // adjust position as a center top of filled rectangle
+    ctx.strokeRect(
+      -0.5 * this.w,
+      this.h - this.fillRatio * this.h,
+      this.w,
+      -this.strokeRatio * this.h,
+    );
+    ctx.fillRect(-0.5 * this.w, 0, this.w, this.fillRatio * this.h);
+    ctx.restore();
+  }
+}
+
+const source = new DoubleRect(50, 450, 50, 50, 0, 0, 1);
+
+const splitter1 = new DoubleRect(50, 350, 50, 10, 135, 2, 1);
+const splitter2 = new DoubleRect(250, 150, 50, 10, 135, 2, 1);
+
+const mirror1 = new DoubleRect(50, 150, 50, 10, 135, 1, 1);
+const mirror2 = new DoubleRect(250, 350, 50, 10, -45, 1, 1);
+
+function MachZehnderCanvas({ size: [width, height], refractionIndex }) {
```

Comment:
for the props validation warning, you can fix it with here 
https://ja.reactjs.org/docs/typechecking-with-proptypes.html